### PR TITLE
refactor(ir): generalize handling and traversal of node sequences

### PR DESCRIPTION
**Pull request by kszucs**
depends on #4421

### CODE REVIEW

File: ibis/expr/rules.py [hunk 1]
```diff
@@ -90,22 +97,17 @@ def just(arg):
     return lambda **_: arg
 
 
+# TODO(kszucs): perhaps rename to nodes_of
 @rule
-def value_list_of(inner, arg, **kwargs):
-    # TODO(kszucs): would be nice to remove ops.ValueList
+def list_of(inner, arg, **kwargs):
```
Comment (kszucs) [inline ibis/expr/rules.py:102]: This rule must be used to define a sequence of nodes which automatically traversed by other utilities. We already have a `tuple_of` rule representing a python tuple of arbitrary types so this naming might be a bit misleading, the reason I chose this was that to align with the operation name `ops.List`.

@cpcloud opinions?

File: ibis/backends/base/sql/alchemy/query_builder.py [hunk 2]
```diff
@@ -262,7 +262,7 @@ def _add_select(self, table_set):
             return result.select_from(table_set)
 
         # otherwise we expect a single FROM in `result`
-        assert num_froms == 1, f"num_froms == {num_froms:d}"
+        # assert num_froms == 1, f"num_froms == {num_froms:d}"
```
Comment (kszucs) [inline ibis/backends/base/sql/alchemy/query_builder.py:265]: If I keep this assertion then `ibis/backends/tests/test_aggregation.py::test_filter` is failing for certain SQL backends.

File: ibis/expr/operations/core.py [hunk 3]
```diff
@@ -1,21 +1,29 @@
...
 from ibis.expr.rules import Shape
-from ibis.expr.schema import Schema
-from ibis.util import UnnamedMarker, is_iterable
+from ibis.util import UnnamedMarker
 
 
 @public
 class Node(Annotable, Comparable):
+    @property
+    def args(self):
```
Comment (kszucs) [inline ibis/expr/operations/core.py:18]: I'd be up to removing of `args` and `argnames` in a follow-up.

Comment (cpcloud) [reply ibis/expr/operations/core.py:18]: Let's leave this as is for now.

File: ibis/expr/rules.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/rules.py:102]: Can we rename `ops.List` to `ops.NodeList` and `list_of` to `node_list_of`?

File: ibis/backends/base/sql/alchemy/query_builder.py (hunk 2 above)
Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/query_builder.py:265]: Well, if everything else is passing then it's probably fine to remove!

Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/query_builder.py:265]: Can you delete this line?

File: ibis/expr/rules.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/rules.py:102]: Can you rename? I'll merge after that and the comment below on removing the assertion.

File: ibis/backends/base/sql/alchemy/query_builder.py (hunk 2 above)
Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/query_builder.py:265]: Actually, this is broken.

Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/query_builder.py:265]: Ah, I think we need to move the assertion. One sec, I'll fix it up.

Comment (kszucs) [reply ibis/backends/base/sql/alchemy/query_builder.py]: Thanks!

File: ibis/expr/rules.py (hunk 1 above)
Comment (kszucs) [reply ibis/expr/rules.py:102]: Sure

**Review by cpcloud — COMMENTED** [review]
Nice work!

**Review by cpcloud — APPROVED** [review]

**Pull request merged**